app/code/Magento/QuoteGraphQl/Model/Cart/AddProductsToCart.php

```diff
@@ -44,23 +44,26 @@ public function __construct(
      *
      * @param Quote $cart
      * @param array $cartItems
+     * @return \Magento\Framework\GraphQl\Exception\GraphQlCartInputException
      * @throws GraphQlInputException
      * @throws \Magento\Framework\Exception\LocalizedException
      * @throws \Magento\Framework\GraphQl\Exception\GraphQlNoSuchEntityException
      */
-    public function execute(Quote $cart, array $cartItems): void
+    public function execute(Quote $cart, array $cartItems): \Magento\Framework\GraphQl\Exception\GraphQlCartInputException
     {
         foreach ($cartItems as $cartItemData) {
             $this->addProductToCart->execute($cart, $cartItemData);
         }
 
         if ($cart->getData('has_error')) {
-            $e = new GraphQlInputException(__('Shopping cart errors'));
+            $e = new \Magento\Framework\GraphQl\Exception\GraphQlCartInputException(__('Shopping cart errors'));
             $errors = $cart->getErrors();
             foreach ($errors as $error) {
                 /** @var MessageInterface $error */
                 $e->addError(new GraphQlInputException(__($error->getText())));
             }
+            $e->addData($cartItems);
```

Comment:
This solution looks like a hack to me. We cannot add regular cart data to the exception.

Unfortunately, I don't have answer off the top of my head how to fix this right away, but this is not the right way to do this. Framework should allow returning both - `data` and `errors` and have predictable and behaviour, since other developer might use this code as well.

Reply:
Totally agree. I didn't mean to push the exact solution. I don't know how to represent the solution in the right way. It is should be easily extendable and not obligatory I guess.

Reply:
I might help someone more involved on GraphQl with more extended knowledge to resolve the issue or at least figure out the solution somehow. I actually have no idea how to push cartData into $result. After all the RESULT should be here \Magento\Framework\GraphQl\Query\QueryProcessor::process() this is the method that returns completed error and data with empty array.

Reply:
Could push something in one of this variables that would be available in the queryProcessor where the result will be handled? 

`\Magento\Framework\GraphQl\Query\QueryProcessor::process`
```
$result  = \GraphQL\GraphQL::executeQuery(
            $schema,
            $source,
            $rootValue,
            $contextValue,
            $variableValues,
            $operationName
        );
```